stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/message/attachment/UploadAttachmentsWorker.kt

```diff
@@ -1,24 +1,51 @@
 package io.getstream.chat.android.offline.message.attachment
 
 import android.content.Context
+import io.getstream.chat.android.client.ChatClient
 import io.getstream.chat.android.client.errors.ChatError
 import io.getstream.chat.android.client.models.Attachment
 import io.getstream.chat.android.client.utils.Result
-import io.getstream.chat.android.offline.ChatDomain
 import io.getstream.chat.android.offline.ChatDomainImpl
 
 internal class UploadAttachmentsWorker(private val appContext: Context) {
-    suspend fun uploadAttachmentsForMessage(channelType: String, channelId: String, messageId: String): Result<Unit> {
+    suspend fun uploadAttachmentsForMessage(
+        channelType: String,
+        channelId: String,
+        messageId: String,
+        chatDomain: ChatDomainImpl,
+        chatClient: ChatClient
+    ): Result<Unit> {
         return try {
-            val domainImpl = (ChatDomain.instance() as ChatDomainImpl)
-            val message = domainImpl.repos.selectMessage(messageId)!!
-            val attachments = domainImpl.channel(channelType, channelId)
-                .uploadAttachments(message)
+            val domainImpl = chatDomain.apply {
+                if (user.value == null) {
+                    if (!chatClient.containsStoredCredentials()) {
+                        return Result.error(ChatError("Could not set user"))
+                    }
+
+                    chatClient.setUserWithoutConnectingIfNeeded()
+                }
+            }
 
-            if (attachments.all { it.uploadState == Attachment.UploadState.Success }) {
+            val message = domainImpl.repos.selectMessage(messageId)
+
+            if (message == null) {
                 Result.success(Unit)
             } else {
-                Result.error(ChatError())
+                val hasPendingAttachment = message.attachments.any { attachment ->
+                    attachment.uploadState == Attachment.UploadState.InProgress
+                }
+
+                if (!hasPendingAttachment) {
+                    return Result.success(Unit)
+                }
+
```

Comment:
Seems like here we should insert message with updated attachments (their `uploadState` should be `InProgress`) for attachments that haven't been sent successfully yet

Reply:
I see that already happens in the `ChannelController.uploadAttachments`. It calls `domainImpl.repos.insertMessage(message)` after the upload happens. Then the message will be updated with the new status of its attachments

Reply:
But just before sending we don't update their state. And consider we're opening app, `connectionRecovered` is triggered, we're checking for messages that wait for attachments and filter out those who's attachment have `InProgress` state, but we don't have such, then we send request to upload again. Theoretically we can have double sending job

Reply:
Discussed it. It's already done in `MessageSendingService`

Reply:
documentation: We talked about that in Slack